### ISSUE TITLE
Declaring multiple outputs in one method

### ISSUE THREAD

**Issue by philippjfr** [pyviz/param#310]
In https://github.com/ioam/param/pull/299 we introduced the ``@output`` decorator after quite a lengthy discussion on how to declare an outputs. Overall the design works reasonably well but during the discussion we rejected the idea of allowing a method with multiple outputs, which after several users have been using the approach for developing apps has I think turned out to be a severe limitation.

Let us illustrate the limitation using a simple example:

```python
class SomeExample(param.Parameterized):

    @param.output()
    def output1(self):
        return self.run()[0]

    @param.output()
    def output2(self):
        return self.run()[1]

    def run(self):
        ... # Some expensive computation  
        return (a, b)
```

We want to declare two outputs for this class, which are both obtained by calling the ``run`` method, but because calling it is expensive we would like to avoid having to call the method twice. Currently this leaves us with only a few options:

1) Implement some ad-hoc caching scheme for ``run``
2) Rely on some inbuilt memoization that param could provide
3) Return both outputs under a single name and then unpack them in the object that's consuming them

All three options are, imo, very suboptimal. 1) Reimplementing a caching scheme for a common problem has too much overhead, 2) memoization would work but seems like overkill for a common scenario and 3) requiring the user to adapt downstream classes to accept the outputs in some nested form only to then unpack them is ugly and inflexible, especially if the objects are themselves complex parameterized objects with meaningful names.

Therefore I would like to suggest that we extend the output decorator to declare the return of multiple objects, e.g. in the above scenario I would suggest something like:

```python
class SomeExample(param.Parameterized):

    @param.output(output1=param.Parameter, output2=param.Parameter)
    def run(self):
        ... # Some expensive computation  
        return {'output1': a, 'output2': b}
```

Declaring multiple return values would **require** you to explicitly declare them in the ``output`` signature and they must match the names provided in the dictionary being returned. I believe this would provide a significant improvement in usability, and since you have to be completely explicit about the outputs being returned there is no ambiguity involved.

Cc: @kcpevey

**Comment by kcpevey** [comment]
I like your suggestion. That's along the same lines I was thinking.

**Comment by jbednar** [comment]
Sounds good to me.

**Comment by philippjfr** [comment]
A modified version of this was implemented in #312, it's now available in param 1.8.2 and I'll make a PR adding support for it to panel pipelines shortly.
